In [1]:
# ============================================================
# RetailPulse | Phase 1: Data Inspection
# Author: Naisha
# Date: May 2026
# Purpose: Load all tables, inspect shape, nulls, and dtypes
# ============================================================

import pandas as pd
import os

# ── 1. Load all tables ──────────────────────────────────────
RAW = "../data/raw/"

tables = {
    "customers"   : "olist_customers_dataset.csv",
    "orders"      : "olist_orders_dataset.csv",
    "order_items" : "olist_order_items_dataset.csv",
    "payments"    : "olist_order_payments_dataset.csv",
    "products"    : "olist_products_dataset.csv",
    "sellers"     : "olist_sellers_dataset.csv",
    "reviews"     : "olist_order_reviews_dataset.csv",
    "geolocation" : "olist_geolocation_dataset.csv",
    "categories"  : "product_category_name_translation.csv"
}

dfs = {}
for name, file in tables.items():
    dfs[name] = pd.read_csv(os.path.join(RAW, file))
    print(f"✓ Loaded {name:15s} → {dfs[name].shape}")

✓ Loaded customers       → (99441, 5)
✓ Loaded orders          → (99441, 8)
✓ Loaded order_items     → (112650, 7)
✓ Loaded payments        → (103886, 5)
✓ Loaded products        → (32951, 9)
✓ Loaded sellers         → (3095, 4)
✓ Loaded reviews         → (99224, 7)
✓ Loaded geolocation     → (1000163, 5)
✓ Loaded categories      → (71, 2)


In [2]:
# ── 2. Health report on every table ─────────────────────────
for name, df in dfs.items():
    print(f"\n{'='*55}")
    print(f" TABLE: {name.upper()}")
    print(f"{'='*55}")
    print(f" Rows: {df.shape[0]:,}  |  Columns: {df.shape[1]}")
    print(f"\n Null counts:")
    nulls = df.isnull().sum()
    nulls = nulls[nulls > 0]
    if len(nulls) == 0:
        print("  → No nulls found ✓")
    else:
        for col, count in nulls.items():
            pct = count / len(df) * 100
            print(f"  → {col}: {count:,} nulls ({pct:.1f}%)")
    print(f"\n Column types:")
    for col, dtype in df.dtypes.items():
        print(f"  → {col}: {dtype}")
        
        # ── 3. Preview each table ────────────────────────────────────
for name, df in dfs.items():
    print(f"\n{'='*55}")
    print(f" TABLE: {name.upper()} — first 3 rows")
    print(f"{'='*55}")
    display(df.head(3))


 TABLE: CUSTOMERS
 Rows: 99,441  |  Columns: 5

 Null counts:
  → No nulls found ✓

 Column types:
  → customer_id: object
  → customer_unique_id: object
  → customer_zip_code_prefix: int64
  → customer_city: object
  → customer_state: object

 TABLE: ORDERS
 Rows: 99,441  |  Columns: 8

 Null counts:
  → order_approved_at: 160 nulls (0.2%)
  → order_delivered_carrier_date: 1,783 nulls (1.8%)
  → order_delivered_customer_date: 2,965 nulls (3.0%)

 Column types:
  → order_id: object
  → customer_id: object
  → order_status: object
  → order_purchase_timestamp: object
  → order_approved_at: object
  → order_delivered_carrier_date: object
  → order_delivered_customer_date: object
  → order_estimated_delivery_date: object

 TABLE: ORDER_ITEMS
 Rows: 112,650  |  Columns: 7

 Null counts:
  → No nulls found ✓

 Column types:
  → order_id: object
  → order_item_id: int64
  → product_id: object
  → seller_id: object
  → shipping_limit_date: object
  → price: float64
  → freight_value: float64

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP



 TABLE: ORDERS — first 3 rows


,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00



 TABLE: ORDER_ITEMS — first 3 rows


,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.9,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.9,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.0,17.87



 TABLE: PAYMENTS — first 3 rows


,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71



 TABLE: PRODUCTS — first 3 rows


,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0



 TABLE: SELLERS — first 3 rows


,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ



 TABLE: REVIEWS — first 3 rows


,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24



 TABLE: GEOLOCATION — first 3 rows


,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP



 TABLE: CATEGORIES — first 3 rows


,product_category_name,product_category_name_english
0,beleza_saude,health_beauty
1,informatica_acessorios,computers_accessories
2,automotivo,auto


In [3]:
dfs['customers'].head()


,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [4]:
dfs['customers']['customer_state'].value_counts()

customer_state
SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: count, dtype: int64

## Customers Table — First Observations

- 5 columns: two IDs, zip code, city, state
- customer_unique_id is the true loyalty identifier
- 42% of customers are from São Paulo (SP)
- Top 3 states cover 67% of all customers
- 14+ states are significantly underserved

In [5]:
dfs['orders'].head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [6]:
dfs['orders']['order_status'].value_counts()

order_status
delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: count, dtype: int64

## Orders Table — First Observations

- 8 columns including purchase, approval, and delivery timestamps
- 97% of orders were successfully delivered
- 625 cancelled orders — needs investigation in cleaning phase
- Can calculate delivery delay = actual vs estimated delivery date

In [7]:
dfs['payments'].head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [8]:
dfs['payments']['payment_type'].value_counts()

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

In [9]:
dfs['payments']['payment_installments'].describe()

count    103886.000000
mean          2.853349
std           2.687051
min           0.000000
25%           1.000000
50%           1.000000
75%           4.000000
max          24.000000
Name: payment_installments, dtype: float64

## Payments Table — First Observations

- payment_value is the core revenue column
- 74% of orders paid by credit card
- 50% of customers pay in full (1 installment)
- Max installments = 24 months — high value purchases
- 3 rows with payment_type = not_defined → drop in cleaning
- Min installments = 0 → data error, fix in cleaning

## Null Report — Key Findings

### Issues to fix in cleaning:
1. Orders: 5 date columns loaded as object → convert to datetime
2. Orders: delivery nulls expected (cancelled/in-transit orders)
3. Products: 610 nulls in category → fill with "unknown"
4. Products: typo in column name → rename to product_name_length
5. Payments: 3 rows with payment_type = not_defined → drop
6. Payments: rows with 0 installments → investigate and drop

### Tables that are clean:
- customers, order_items, sellers, geolocation, categories